### ISSUE TITLE
Inconsistent behaviour between juice() and bake() when using step_filter()

### ISSUE THREAD

**Issue by mattwarkentin** [tidymodels/recipes#554]
### Minimal, reproducible example:

Maybe I'm doing something wrong, but it seems like `step_filter()` is just not being applied properly when `bake()`ing compared to `juice()`ing. As you can see, the returned `tibble`s differ in that `juice()` filters in the proper rows, and `bake()` seemingly doesn't do any filtering and returns the input `tibble`.

Any thoughts on what is going on here?

``` r
library(recipes)

recipe(iris) %>% 
  step_filter(Species == "setosa") %>% 
  prep() %>% 
  juice()
#> # A tibble: 50 x 5
#>    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#>           <dbl>       <dbl>        <dbl>       <dbl> <fct>  
#>  1          5.1         3.5          1.4         0.2 setosa 
#>  2          4.9         3            1.4         0.2 setosa 
#>  3          4.7         3.2          1.3         0.2 setosa 
#>  4          4.6         3.1          1.5         0.2 setosa 
#>  5          5           3.6          1.4         0.2 setosa 
#>  6          5.4         3.9          1.7         0.4 setosa 
#>  7          4.6         3.4          1.4         0.3 setosa 
#>  8          5           3.4          1.5         0.2 setosa 
#>  9          4.4         2.9          1.4         0.2 setosa 
#> 10          4.9         3.1          1.5         0.1 setosa 
#> # … with 40 more rows

recipe(iris) %>% 
  step_filter(Species == "setosa") %>% 
  prep() %>% 
  bake(new_data = iris)
#> # A tibble: 150 x 5
#>    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#>           <dbl>       <dbl>        <dbl>       <dbl> <fct>  
#>  1          5.1         3.5          1.4         0.2 setosa 
#>  2          4.9         3            1.4         0.2 setosa 
#>  3          4.7         3.2          1.3         0.2 setosa 
#>  4          4.6         3.1          1.5         0.2 setosa 
#>  5          5           3.6          1.4         0.2 setosa 
#>  6          5.4         3.9          1.7         0.4 setosa 
#>  7          4.6         3.4          1.4         0.3 setosa 
#>  8          5           3.4          1.5         0.2 setosa 
#>  9          4.4         2.9          1.4         0.2 setosa 
#> 10          4.9         3.1          1.5         0.1 setosa 
#> # … with 140 more rows
```

<sup>Created on 2020-08-17 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by EmilHvitfeldt** [comment]
`step_filter()` has the `skip` argument set to `TRUE` by default. The `skip` argument determines if the step should be skipped when the recipe is `bake`d.

Care should be taken when using `skip = FALSE`; in most instances that affect the rows of the data being predicted, this step probably should not be applied.

``` r
library(recipes)

recipe(iris) %>% 
  step_filter(Species == "setosa", skip = FALSE) %>% 
  prep() %>% 
  bake(new_data = iris)
#> # A tibble: 50 x 5
#>    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
#>           <dbl>       <dbl>        <dbl>       <dbl> <fct>  
#>  1          5.1         3.5          1.4         0.2 setosa 
#>  2          4.9         3            1.4         0.2 setosa 
#>  3          4.7         3.2          1.3         0.2 setosa 
#>  4          4.6         3.1          1.5         0.2 setosa 
#>  5          5           3.6          1.4         0.2 setosa 
#>  6          5.4         3.9          1.7         0.4 setosa 
#>  7          4.6         3.4          1.4         0.3 setosa 
#>  8          5           3.4          1.5         0.2 setosa 
#>  9          4.4         2.9          1.4         0.2 setosa 
#> 10          4.9         3.1          1.5         0.1 setosa 
#> # … with 40 more rows
```

<sup>Created on 2020-08-17 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by mattwarkentin** [comment]
Thanks for the response, Emil. Could you please expand on what you meant by the text below?

> ...in most instances that affect the rows of the data being predicted, this step probably should not be applied.

**Comment by EmilHvitfeldt** [comment]
Lets us say you are trying to build a model. We start with splitting the data into testing and training

``` r
library(tidymodels)
set.seed(1234)
mtcars_split <- initial_split(mtcars)

mtcars_split
#> <Analysis/Assess/Total>
#> <24/8/32>

mtcars_training <- training(mtcars_split)
mtcars_testing <- testing(mtcars_split)
```

Notice how we have 24 training observations and 8 testing observations.

We now specify a preprocessor and fit a linear model. We use `step_filter()` with `skip = FALSE` because we needed to remove some irregular observations from our training set.

```r
rec_spec <- recipe(mpg ~ ., data = mtcars_training) %>%
  step_filter(disp <= 400, skip = FALSE) %>%
  step_normalize(all_predictors()) %>%
  prep()

model_fit <- linear_reg() %>%
  set_engine("lm") %>%
  fit(mpg ~ ., data = juice(rec_spec))
```

predicting on our testing set gives us back 6 predictions. 

```r
predict(model_fit, bake(rec_spec, mtcars_testing))
#> # A tibble: 6 x 1
#>   .pred
#>   <dbl>
#> 1  18.0
#> 2  34.3
#> 3  12.3
#> 4  14.5
#> 5  27.2
#> 6  23.7
```

It is generally unlikely that you want this behavior in your modeling workflow. We tend to want the number of predicted values to be the same as the number of rows in our testing set. Setting `skip = FALSE` in steps that could modify the number of rows is generally not advised when doing predictive modeling.

**Comment by juliasilge** [comment]
[This might be a more clear way to put it](https://www.tidyverse.org/blog/2020/06/recipes-0-1-13/#general-changes):

> First, `step_filter()`, `step_slice()`, `step_sample()`, and `step_naomit()` had their defaults for `skip` changed to `TRUE`. In the vast majority of applications, these steps should not be applied to the test or assessment sets. 

These are all row-filtering operations and typically when you are building a model, you _do_ want a prediction for every row in the test set. Compare [this default value for `skip`](https://recipes.tidymodels.org/reference/step_filter.html) to, say, the [one for `step_other()`](https://recipes.tidymodels.org/reference/step_other.html).

If you are in a situation where you actually don't want to `skip`, then you can specify `skip = FALSE` in that recipe step.

**Comment by mattwarkentin** [comment]
Thank you both for your responses. Totally makes sense!
